res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,3061 @@
+        }
+        if (this.mixer.firstPressedFxSelector === this.number) {
+            this.mixer.firstPressedFxSelector = null;
+            this.mixer.resetFxSelectorColors();
+        }
+        if (this.mixer.secondPressedFxSelector !== null) {
+            this.mixer.comboSelected = true;
+        }
+        this.mixer.secondPressedFxSelector = null;
+    }
+
+}
+
+
+class QuickEffectButton extends Button {
+    constructor(options) {
+        super(options);
+        if (this.mixer === undefined) {
+            throw Error("The mixer must be specified");
+        }
+        if (this.number === undefined || !Number.isInteger(this.number) || this.number < 1) {
+            throw Error("number attribute must be an integer >= 1");
+        }
+        this.group = "[QuickEffectRack1_[Channel" + this.number + "]]";
+        this.outConnect();
+    }
+    onShortPress() {
+        if (this.mixer.firstPressedFxSelector === null) {
+            script.toggleControl(this.group, "enabled");
+        } else {
+            const presetNumber = this.mixer.calculatePresetNumber();
+            this.color = QuickEffectPresetColors[presetNumber - 1];
+            engine.setValue(this.group, "loaded_chain_preset", presetNumber + 1);
+            this.mixer.firstPressedFxSelector = null;
+            this.mixer.secondPressedFxSelector = null;
+            this.mixer.resetFxSelectorColors();
+        }
+    }
+    onLongRelease() {
+        if (this.mixer.firstPressedFxSelector === null) {
+            script.toggleControl(this.group, "enabled");
+        }
+    }
+    output(enabled) {
+        if (enabled) {
+            this.send(this.color + this.brightnessOn);
+        } else {
+            // It is easy to mistake the dim state for the bright state, so turn
+            // the LED fully off.
+            this.send(this.color + this.brightnessOff);
+        }
+    }
+    presetLoaded(presetNumber) {
+        this.color = QuickEffectPresetColors[presetNumber - 2];
+        this.outConnections[1].trigger();
+    }
+    outConnect() {
+        if (this.group !== undefined) {
+            const connection0 = engine.makeConnection(this.group, "loaded_chain_preset", this.presetLoaded.bind(this));
+            if (connection0) {
+                this.outConnections[0] = connection0;
+            } else {
+                console.warn("Unable to connect '" + this.group + ".loaded_chain_preset' to the controller output. The control appears to be unavailable.");
+            }
+            const connection1 = engine.makeConnection(this.group, "enabled", this.output.bind(this));
+            if (connection1) {
+                this.outConnections[1] = connection1;
+            } else {
+                console.warn("Unable to connect '" + this.group + ".enabled' to the controller output. The control appears to be unavailable.");
+            }
+        }
+    }
+}
+
+/*
+ * Kontrol S4 Mk3 hardware-specific constants
+ */
+
+Pot.prototype.max = 2 ** 12 - 1;
```

Comment:
How can I find that?

Reply:
Record the USB traffic when you execute the Native Instrument calibration procedure: https://support.native-instruments.com/hc/en-us/articles/360017006117-How-to-Calibrate-the-TRAKTOR-S4-MK3-Controls
and find out, if Traktor Pro reads the calibration data at startup or if the calculation is done inside the controller.

For reference, there is an unfinished PR for the S2 Mk1, which implements such a functionality: https://github.com/mixxxdj/mixxx/pull/11237
Since support for HID FeatureReports is new in 2.4, no other mapping does this yet.

Reply:
I cannot see any additional reports being read at the startup. When doing the procedure, it seems to set value on the report `0xf4` (ask to put all Pots to zero > press OK > value is set on `0xf4`, then again with pots at max and pots at middle). Perhaps the value is normalized by the device?
It could also be in the report 0x3: we are only interpreting 12 bytes from the 70 that get sent, it could be that some of these contain calibration values. However, beside purposefully messing up my device calibration, I'm not quite sure how I could do that, and I am not fully confident on breaking the calibration on purpose...